bundles/core/org.eclipse.smarthome.core.thing.test/src/test/java/org/eclipse/smarthome/core/thing/firmware/FirmwareUpdateServiceTest.java

```diff
@@ -647,7 +647,9 @@ public void testEvents() {
         events.get().addAll(eventCaptor.getAllValues());
         List<Event> list = events.get().stream().filter(event -> event instanceof FirmwareUpdateProgressInfoEvent)
                 .collect(Collectors.toList());
-        assertTrue(list.size() >= SEQUENCE.length);
+        waitForAssert(() -> {
```

Comment:
I doubt that this will fix the test. Here you are waiting in case the event list does not have the correct size. But there is no background thread which does modify the list during waiting.
My assumption is that lines 644 - 649 could be the problem: the waitForAssert in line 644 verifies atLeast SEQUENCE.length times, and later (line 648) some of the captured events might be filtered out. If this is the case then the list will be shorter than SEQUENCE.length. wdyt?

Reply:
My assumption is that at line 639 firmwareUpdateService.getFirmwareStatusInfo() calls internally eventPublisher.post() which posts FirmwareStatusInfoEvent. At line 644 we are waiting for four events, but we don't filter their types. This way the first four events are one FirmwareStatusInfoEvent and three FirmwareUpdateProgressInfoEvent.

Reply:
yes, this seems to be correct. Verifying on `SEQUENCE.length + 1` invocations of `mockPublisher#post` might do it then.

Reply:
PR updated regarding htreu's last comment.